File: Annexe9/app/src/main/java/com/example/annexe9/MainActivity.java
package com.example.annexe9;

import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    // Déclaration des TextViews pour afficher les réponses
    TextView reponse1, reponse2, reponse3, reponse4, reponse5;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Configuration de l'affichage
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        // Initialisation des TextViews
        reponse1 = findViewById(R.id.reponse1);
        reponse2 = findViewById(R.id.reponse2);
        reponse3 = findViewById(R.id.reponse3);
        reponse4 = findViewById(R.id.reponse4);
        reponse5 = findViewById(R.id.reponse5);

        // URL du JSON en ligne
        String url = "https://api.jsonbin.io/v3/b/6733b233ad19ca34f8c9149a?meta=false";

        // Appel de la méthode pour effectuer une requête avec Volley
        fetchJsonWithVolley(url);

        // Lecture du fichier JSON local et traitement avec GSON
        processLocalJson();
    }

    /**
     * Méthode pour effectuer une requête JSON en ligne avec Volley.
     */
    private void fetchJsonWithVolley(String url) {
        RequestQueue queue = Volley.newRequestQueue(this);

        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(
                Request.Method.GET, url, null,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        try {
                            // Extraction et affichage des données
                            JSONObject menu = response.getJSONObject("menu");
                            String header = menu.getString("header");
                            Toast.makeText(getApplicationContext(), "Response: " + response.toString(), Toast.LENGTH_LONG).show();
                            // Affichage dans les TextViews
                            reponse1.setText(header);
                        } catch (Exception e) {
                            e.printStackTrace();
                            Toast.makeText(MainActivity.this, "Erreur : " + e.getMessage(), Toast.LENGTH_LONG).show();
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Toast.makeText(MainActivity.this, "Erreur de requête : " + error.getMessage(), Toast.LENGTH_LONG).show();
                    }
                }
        );

        queue.add(jsonObjectRequest);
    }

    /**
     * Méthode pour lire un fichier JSON local et le traiter avec GSON.
     */
    private void processLocalJson() {
        // Lecture du fichier JSON local
        String jsonString = readLocalJson();

        // Traitement du JSON avec GSON
        if (jsonString != null) {
            Gson gson = new Gson();
            Type listType = new TypeToken<List<Voiture>>() {}.getType();
            List<Voiture> voitures = gson.fromJson(jsonString, listType);

            // Calcul et affichage de la moyenne des prix
            int total = 0;
            for (Voiture voiture : voitures) {
                total += voiture.getPrix();
            }
            float moyenne = (float) total / voitures.size();
            reponse2.setText("Prix moyen : " + moyenne);
        }
    }

    /**
     * Méthode pour lire le contenu d'un fichier JSON local.
     */
    private String readLocalJson() {
        InputStream is = getResources().openRawResource(R.raw.voiture);
        BufferedReader reader = new BufferedReader(new InputStreamReader(is));
        StringBuilder jsonBuilder = new StringBuilder();
        String line;

        try {
            while ((line = reader.readLine()) != null) {
                jsonBuilder.append(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(this, "Erreur de lecture du fichier JSON", Toast.LENGTH_LONG).show();
            return null;
        } finally {
            try {
                reader.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return jsonBuilder.toString();
    }
}

File: tp2/app/src/main/java/com/example/annexe1/MainActivity.kt
package com.example.annexe1

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.Vector

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        retourner()
        signer()
    }

    fun retournerLigneProf(): Int{
        val fis : FileInputStream = openFileInput("fichier_annexe1.txt")
        val isr = InputStreamReader(fis)
        val br = BufferedReader(isr)
        var compteur = 0

        while (br.readLine() != null){
            compteur++
        }

        br.close()
        return compteur
    }

    fun retournerCharProf(): Int{
        var compteurChar = 0
        val fis : FileInputStream = openFileInput("fichier_annexe1.txt")
        val isr = InputStreamReader(fis)
        val br = BufferedReader(isr)
        var ligne = br.readLine()

        while (ligne != null){
            compteurChar += ligne.length
            ligne = br.readLine()
        }

        br.close()
        return compteurChar
    }


    fun nombreDeCProf() : Int {
        var compteurChar = 0
        val fis : FileInputStream = openFileInput("fichier_annexe1.txt")
        val isr = InputStreamReader(fis)
        val br = BufferedReader(isr)
        var c:Int = br.read()
        while(c != -1){
            if (c.toChar() == 'c' || c.toChar() == 'C')
                compteurChar ++
            c = br.read()
        }
        br.close()
        return compteurChar
    }

    // la ligne (le nom) est passé en paramètre
    fun signerProf(ligne:String){
        val fos : FileOutputStream = openFileOutput("fichier_annexe1.txt", MODE_APPEND)
        val osw = OutputStreamWriter(fos)
        val bw = BufferedWriter(osw)

        bw.newLine();
        bw.write(ligne)

        bw.close()
    }

    fun retourner() {
        var char = 0
        var c = 0
        var ligne = 0
        val fis : FileInputStream = openFileInput("fichier_annexe1.txt")
        val isr = InputStreamReader(fis)
        val br = BufferedReader(isr)


        br.forEachLine { line ->
            ligne += 1
            char += line.filter { it != ' ' }.length
            c += line.filter { it == 'c' || it == 'C' }.length
        }


        println("nombre de lignes : $ligne")
        println("nombre de characteres : $char")
        println("nombre de 'C' et 'c' : $c")

        br.close()
        return
    }


    fun signer(){

        // val bwOptimisé : BufferedWriter = openFileOutput("fichier_annexe1.txt", MODE_APPEND).bufferedWriter()

        val nom = "Edouard"
        val fos : FileOutputStream = openFileOutput("fichier_annexe1.txt", MODE_APPEND)
        val osw = OutputStreamWriter(fos)
        val bw = BufferedWriter(osw)



        bw.write("\n$nom")

        bw.close()

    }

}